Make the cluster-side DNS resolver configurable

thallgren opened the pull request:
## Description

This PR contains improvements that enables configuration of the local and remote IP address of the DNS resolver, and limit the number of `LookupHost()` requests that the client sends to the cluster. The configuration is an extension to the kubeconfig's Cluster definition. The improvements are mainly in five areas.

1. The IP of the local resolver can now be set in the cluster configuration and will become the default unless `--dns` is passed on the command line. This setting is only relevant to Linux workstations that aren't configured to use `systemd.resolved`.
2. The IP of the remote resolver can be configured. This can be useful for systems for which Telepresence doesn't have automatic DNS discovery.
3. The overriding DNS resolver, used on Linux without `systemd-resolved` and in Docker containers, receives all DNS requests that are issued on the client. This is very different from the other resolvers which configure the search paths, domains, and routes, that they want to handle. Since we don't want to send all DNS requests over to the cluster, this resolver now checks if the last name in a QN is a namespace, or if the QN ends with ".svc" or ".cluster.local". When this is the case, the name will not be subjected to concatenations with the search path.
4. The DNS resolvers on all platforms are now configured with an `exclude-suffixes` list by default. This list contains a set of very well known top level domains such as ".com", ".org", and a few more. QNs that have those suffixes will not be dealt with by the resolver. Instead it will return not found, or in case of the overriding resolver, use the fallback. The list is configurable under the new `DNS` heading in the Telepresence config file.
5. There's also an `include-suffixes` list. This list has higher priority than the `exclude-suffixes` list and is intended to be used when it's essential that some suffixes are sent to the cluster resolver that normally wouldn't be. In addition to overriding the `exclude-suffixes`, an entry in this list will also become a "route" in the `systemd-resolved` case and a `domain` on MacOS. This to ensure that the DNS request reaches the resolver in the first place. This solution is primarily intended for deployments that use custom DNS configurations, and should be seen as a workaround for the fact that we cannot automatically cater for what domains such configurations may have.

## Checklist

 - [x] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes.
 - [x] My change is adequately tested.

Commit:
Fix proper logging when returning OK but empty for unmatched query type.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Don't apply search path on known namespaces in overriding resolver

If a QN ends with a namespace, it shouldn't be subjected to search path
concatenation. Doing so results in unnecessary `LookupHost()` calls.
The same is true for names that end with ".cluster.local" or ".svc".

This commit also establishes a more consistent handling of the ending
dot.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make DNS resolver configurable

This commit contains an extension for the kubeconfigs `Cluster`
configuration. The extension will allow the DNS to be configured
with the following:

- local-ip: The address of the local DNS resolver. It is overridden by
  the `--dns` command line option.
- remote-ip: The address of the cluster's DNS server. Attempts are made
  to auto-discover this address when not explicitly set.
- exclude-suffixes: Domain name suffixes that the resolver should avoid
  sending to the cluster.
- include-suffixes: Domain name suffixes that the resolver should be
  configured to receive and send to the cluster.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Update CHANGELOG.md

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

thallgren commented on pkg/client/daemon/outbound.go:
```diff
@@ -37,7 +37,7 @@ type outbound struct {
 	// Namespaces, accessible using <service-name>.<namespace-name>
 	namespaces map[string]struct{}
-	domains    map[string]iputil.IPs
+	domains    map[string]struct{}
```
The `domains` map was actually a leftover from 2.2 but is now needed by the MacOS resolver for a slightly different purpose.

donnyyung commented on pkg/client/daemon/outbound.go:
```diff
@@ -187,9 +207,8 @@ func (o *outbound) resolveInCluster(c context.Context, qType uint16, query strin
-	tos := client.GetConfig(c).Timeouts
-	c, cancel := tos.TimeoutContext(c, client.TimeoutTrafficManagerAPI)
+	// Give the cluster lookup a reasonable timeout.
+	c, cancel := context.WithTimeout(c, 4*time.Second)
```
I think we should make this timeout configurable, add a TODO comment saying "we * should * make this configurable at some point", or provide a reason why this should be hardcoded to 4 seconds.

thallgren replied:
That's an oversight. This timeout should definitely be added.

donnyyung commented on pkg/client/connector/k8s_config.go:
```diff
@@ -1,17 +1,47 @@
+	IncludeSuffixes []string `json:"include-suffixes,omitempty"`
+}
+
+// kubeconfigExtension is an extension read from the selected kubeconfig Cluster.
```
Is the reasoning for putting this configuration in the kubeconfig because users could potentially have different values for different clusters?  

I guess I'm torn because I'd prefer for all of our config to be in one place (e.g. `telepresence/config.yml`) because once configuration is spread across multiple files, I feel like it can get confusing to figure out what you configure where.  

I think I still need to do some more thinking about this, but just wanted to check + ensure I understood the reasoning for this decision.

thallgren replied:
I had this in the `config.yml` initially. Then @LukeShu suggested that we should use the kubeconfig extension mechanism for the `also-proxy` stuff because it's bound to be cluster specific and I really liked that idea. Since the DNS resolver resides in the cluster nowadays, it's also cluster specific (besides the `local-ip`, but separating that out seemed worse than keeping the config together). The include-suffixes for instance, is partly aimed at solving the problem with pods using custom DNS-resolvers.

donnyyung replied:
Ah okay that makes a lot of sense. I was on the fence initially, but after thinking more about it and listening to our reasoning, I think it makes more sense for cluster-specific config to be added via a kubeconfig extension and the `config.yml` can be for more "global" telepresence config. I think as long as we document that well, then it shouldn't be confusing for our users.

Commit:
Make timeout for cluster side lookups configurable.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

LukeShu commented on pkg/client/daemon/outbound.go:
```diff
@@ -135,18 +137,40 @@ func (o *outbound) routerServerWorker(c context.Context) (err error) {
 var localhostIPv6 = []net.IP{{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1}}
 var localhostIPv4 = []net.IP{{127, 0, 0, 1}}
 
+func (o *outbound) doClusterLookup(query string) bool {
```
Based on the name of this function, I was expecting it to actually _do_ the lookup.  Perhaps `shouldDoClusterLookup`?

LukeShu commented on pkg/client/daemon/outbound.go:
```diff
@@ -135,18 +137,40 @@ func (o *outbound) routerServerWorker(c context.Context) (err error) {
 var localhostIPv4 = []net.IP{{127, 0, 0, 1}}
 
+func (o *outbound) doClusterLookup(query string) bool {
+	if strings.HasSuffix(query, dotKubernetesZone) && strings.Count(query, ".") < 4 {
```
Why `strings.Count(query, ".") < 4`?

LukeShu replied:
I guess that's sorta answered by the old code
```patch
-	// We don't care about queries to the kubernetes zone unless they have at least two additional elements.
-	if strings.HasSuffix(query, dotKubernetesZone) && strings.Count(query, ".") < 3 {
```
but I still don't know why we don't care unless they have at least two additional elements.

thallgren replied:
To my knowledge there are no names of interest that would be just `xxx.cluster.local`. You would have at least `xxx.svc.cluster.local`, `xxx.pod.cluster.local` or similar.

thallgren replied:
... and we are using `cluster.local` as the domain which means that queries like `xxx.cluster.local` do arrive quite often.

LukeShu commented on pkg/client/daemon/outbound.go:
```diff
@@ -135,18 +137,40 @@ func (o *outbound) routerServerWorker(c context.Context) (err error) {
+
 func (o *outbound) resolveInCluster(c context.Context, qType uint16, query string) []net.IP {
-	query = query[:len(query)-1]
 	query = strings.ToLower(query) // strip of trailing dot
```
I think that `// strip of trailing dot` had been intended for the previous deleted line, and therefore should also be deleted?

thallgren replied:
Yup. That comment should not be there.

LukeShu commented on pkg/client/daemon/outbound_linux.go:
```diff
@@ -36,30 +36,51 @@ func (o *outbound) dnsServerWorker(c context.Context) error {
-		if strings.HasSuffix(q, ".local") {
-			return nil
+// applySearch returns true if search path should be applied
+func (o *outbound) applySearch(query string) bool {
```
Again; confusing name.  Perhaps `shouldApplySearch`?

LukeShu commented on pkg/client/daemon/outbound_linux.go:
```diff
@@ -36,30 +36,51 @@ func (o *outbound) dnsServerWorker(c context.Context) error {
-			}
-			ips = o.resolveInCluster(c, qType, q+s)
-			if len(ips) > 0 {
-				break
+			if ips := o.resolveInCluster(c, qType, query+s); len(ips) > 0 {
+				return ips
 			}
 		}
 	}
-	return ips
+	return o.resolveInCluster(c, qType, query)
```
I don't get why we have to loop over the search path here.  I understood why we _used to_ have to loop over it; in the pre-`resolveInCluster` days.  But why are we still doing this client-side, when we can just trim off the trailing `.` and let the agent apply its search path when responding to the LookupHost request?

thallgren replied:
This logic is only used by the overriding resolver. And yes, in most cases it would work to just pass things over. The exception is for local-only intercepts. It should be also be noted that there's a coming limitation that will limit the search path length to one, namely the intercepted namespace so there will be no repeated search.

thallgren replied:
I'm working on a PR to impose the limitation to only be able to intercept one namespace at a time. In that PR, the use of  a search path  is replaced by passing the optional local-only intercepted namespace.

LukeShu commented on rpc/daemon/daemon.proto:
```diff
@@ -47,6 +47,25 @@ message IPNet {
 }
 
+// DNS configuration for the local DNS resolver
+message DNS {
```
Naming this just `DNS` seems a little... oblique.  Perhaps `DNSConfig`?

LukeShu commented on pkg/client/connector/k8s_config.go:
```diff
@@ -1,17 +1,50 @@
+	// to the IP of the kube-dns.kube-system or the dns-default.openshift-dns service.
+	RemoteIP iputil.IPKey `json:"remote-ip,omitempty"`
+
+	// ExcludeSuffixes are suffixes for which the DNS resolver will always fail (or fallback
+	// in case of the overriding resolver).
+	ExcludeSuffixes []string `json:"exclude-suffixes,omitempty"`
```
What does "fail" mean in this context? Returning NXDOMAIN?

thallgren replied:
Yes, it will return NXDOMAIN.

LukeShu commented on pkg/client/connector/k8s_config.go:
```diff
@@ -1,17 +1,50 @@
 )
 
+// The dns config part of the kubeconfigExtension that is read from the kubeconfig
+type dns struct {
```
Again, unclear name. Perhaps `dnsConfig`?

LukeShu replied:
Actually, `dnsConfig` is almost just as bad.  How about `kubeconfigDNS`?

thallgren replied:
This is a subtype in the`kubeconfigExtension`. I think `kubconfigDNS` in `kubeconfigExtension` is a bit of an overkill. Hence just `dns`.

LukeShu commented on pkg/client/connector/k8s_config.go:
```diff
@@ -21,6 +54,8 @@ type k8sConfig struct {
 	config      *rest.Config
 }
 
+const configExtension = "telepresence.getambassador.io"
```
This should probably be `telepresence.io`? Since it's CNCF and not Ambassador Labs?

thallgren replied:
We use `telpresence.getambassador.io` in other places. The annotation in the workloads for instance, is `telepresence.getambassador.io/actions`. I like the consistency, but perhaps we should consider changing the annotation too?

LukeShu replied:
Hmm, so we are.  I guess as long as that's what we're already doing, then it's good enough.

LukeShu commented on pkg/client/daemon/outbound.go:
```diff
@@ -135,18 +137,40 @@ func (o *outbound) routerServerWorker(c context.Context) (err error) {
+
+	query = query[:len(query)-1] // skip last dot
+
+	// Always include configured includeSuffixes
+	for _, sfx := range o.dnsConfig.IncludeSuffixes {
+		if strings.HasSuffix(query, sfx) {
+			return true
+		}
+	}
+
+	// Skip configured excludeSuffixes
+	for _, sfx := range o.dnsConfig.ExcludeSuffixes {
+		if strings.HasSuffix(query, sfx) {
+			return false
+		}
+	}
```
Is giving `includeSuffixes` precedence over `excludeSuffixes` the right call?

thallgren replied:
The reason for making that call is that there's a list of default `excludeSuffixes`. The idea is that you might want to keep most of them but override a few. The `includeSuffixes` has no default list.

LukeShu commented:
I don't suppose we could get that `DNS` gRPC object to be included in the `telepresence status` output?

LukeShu commented on pkg/client/connector/k8s_config.go:
```diff
@@ -1,17 +1,50 @@
+	IncludeSuffixes []string `json:"include-suffixes,omitempty"`
+
+	// The maximum number of seconds to wait for a cluster side host lookup.
+	LookupTimeout int32 `json:"lookup-timeout,omitempty"`
```
Can we use a [`metav1.Duration`](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#Duration) instead of an int, so that it doesn't need to be a whole-number of seconds?

In the YAML it'd be a string:
```yaml
lookup-timeout: 4s
```

thallgren replied:
I thought about that but I'm concerned that users will forget the 's' and run into problems that may be difficult to understand. Do you think it makes sense to configure this timeout in fractions of seconds?

thallgren replied:
Thought about it some more. Duration it is.

LukeShu reviewed (changes requested):
On the whole: Looks good, yay!

I do have some concerns, though.  From highest priority to lowest priority:

- Change Request: I think it should be using `telepresence.io` instead of `telepresence.getambassador.io`, since it's a CNCF project not an Ambassador Labs project.
- Change Request: Use `metav1.Duration` to configure the timeout, rather than an `int` number of seconds.
- Question: I don't get why we still need to loop over the search-path client-side.
- Question: One of the comments mentions "always fail", but it's not obvious what exactly "fail" means here.  That should be clarified in the comment. ("always return NXDOMAIN"?)
- Question: It is unclear to my why we don't send `SINGLE.cluster.local` queries to the cluster and only send them if there are multiple lables before the `.cluster.local` (other than "that's what it's always done")
- Nitpick: There are some confusing names that could be improved.
- Nitpick: The obsolete `// strip of trailing dot` comment should be deleted.

Probably not in this PR (or in 2.3.1), but if you're up for it:

- Feature request: Include the `DNS` config object in the `telepresence status` output.

Commit:
Changes in response to review.

- Rename dns struct to dnsConfig and message DNS to DNSConfig
- Change wording about failing to "return NXDOMAIN"
- Use metav1.Duration instead of seconds for timeout
- Rename extension telepresence.getambassador.io -> telepresence.io
- Prepend confusing function names with "should"
- Remove bad "trim trailing dot" comment
- Add comment to explain why we need `resolveInSearch()`

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

LukeShu reviewed (approved):
Thanks for making those changes. Hit it!
